File: ai-gateway-go/internal/dto/serializers.go
package dto

import (
	"time"

	"github.com/porsche/ai-gateway-go/internal/models"
)

func FormatTime(t time.Time) string {
	if t.IsZero() {
		return time.Now().UTC().Format(time.RFC3339Nano)
	}
	if t.Location() == time.UTC {
		return t.Format(time.RFC3339Nano)
	}
	return t.UTC().Format(time.RFC3339Nano)
}

func UserProfile(user *models.User) map[string]interface{} {
	return map[string]interface{}{
		"id":                user.ID,
		"phone":             user.Phone,
		"nickname":          user.Nickname,
		"is_verified":       user.IsVerified,
		"plan_type":         string(user.PlanType),
		"total_tokens_used": user.TotalTokensUsed,
		"dataset_calls":     user.DatasetCalls,
		"daily_calls_used":  user.DailyCallsUsed,
		"daily_call_limit":  user.DailyCallLimit,
		"created_at":        FormatTime(user.CreatedAt),
	}
}

func Message(msg models.Message) map[string]interface{} {
	return map[string]interface{}{
		"id":                  msg.ID,
		"role":                msg.Role,
		"content":             msg.Content,
		"model":               msg.Model,
		"dataset_used":        msg.DatasetUsed,
		"dataset_attribution": msg.DatasetAttribution,
		"tokens":              msg.Tokens,
		"created_at":          FormatTime(msg.CreatedAt),
	}
}

func Conversation(conv *models.Conversation, includeMessages bool) map[string]interface{} {
	out := map[string]interface{}{
		"id":              conv.ID,
		"title":           conv.Title,
		"model":           conv.Model,
		"dataset_enabled": conv.DatasetEnabled,
		"dataset_ids":     conv.DatasetIDs,
		"created_at":      FormatTime(conv.CreatedAt),
		"updated_at":      FormatTime(conv.UpdatedAt),
	}
	if includeMessages {
		msgs := make([]map[string]interface{}, 0, len(conv.Messages))
		for _, m := range conv.Messages {
			msgs = append(msgs, Message(m))
		}
		out["messages"] = msgs
	}
	return out
}

func Dataset(ds *models.Dataset) map[string]interface{} {
	return map[string]interface{}{
		"id":              ds.ID,
		"name":            ds.Name,
		"slug":            ds.Slug,
		"category":        string(ds.Category),
		"description":     ds.Description,
		"status":          string(ds.Status),
		"current_version": ds.CurrentVersion,
		"token_count":     ds.TokenCount,
		"vector_status":   string(ds.VectorStatus),
		"access_plans":    ds.AccessPlans,
		"created_at":      FormatTime(ds.CreatedAt),
	}
}

func Order(o *models.Order) map[string]interface{} {
	out := map[string]interface{}{
		"id":                o.ID,
		"order_no":          o.OrderNo,
		"plan_type":         string(o.PlanType),
		"amount":            o.Amount,
		"status":            string(o.Status),
		"invoice_requested": o.InvoiceRequested,
		"created_at":        FormatTime(o.CreatedAt),
		"paid_at":           nil,
	}
	if o.PaidAt != nil {
		out["paid_at"] = FormatTime(*o.PaidAt)
	}
	return out
}

func AdminUser(user *models.User) map[string]interface{} {
	return map[string]interface{}{
		"id":                user.ID,
		"phone":             user.Phone,
		"nickname":          user.Nickname,
		"plan_type":         string(user.PlanType),
		"status":            string(user.Status),
		"is_verified":       user.IsVerified,
		"total_tokens_used": user.TotalTokensUsed,
		"dataset_calls":     user.DatasetCalls,
		"created_at":        FormatTime(user.CreatedAt),
	}
}

func AuditLog(log *models.AuditLog) map[string]interface{} {
	return map[string]interface{}{
		"id":         log.ID,
		"user_id":    log.UserID,
		"action":     log.Action,
		"resource":   log.Resource,
		"detail":     log.Detail,
		"ip":         log.IP,
		"created_at": FormatTime(log.CreatedAt),
	}
}

func ModelHealth(h *models.ModelHealth) map[string]interface{} {
	out := map[string]interface{}{
		"model_name":     h.ModelName,
		"provider":       h.Provider,
		"is_available":   h.IsAvailable,
		"avg_latency_ms": h.AvgLatencyMs,
		"error_rate":     h.ErrorRate,
		"last_checked_at": nil,
	}
	if h.LastCheckedAt != nil {
		out["last_checked_at"] = FormatTime(*h.LastCheckedAt)
	}
	return out
}

File: ai-gateway/app/common/schemas/serializers.py
"""ORM → API response builders (avoid Pydantic/from_attributes edge cases)."""

from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy.orm import attributes

from app.repository.enum_utils import enum_value
from app.repository.models import Conversation, Message, User
from app.common.schemas.conversation import ConversationResponse, MessageResponse
from app.common.schemas.user import UserProfileResponse


def _as_utc(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt


def message_response(msg: Message) -> MessageResponse:
    return MessageResponse(
        id=msg.id,
        role=msg.role,
        content=msg.content,
        model=msg.model,
        dataset_used=bool(msg.dataset_used),
        dataset_attribution=msg.dataset_attribution,
        tokens=int(msg.tokens or 0),
        created_at=_as_utc(msg.created_at),
    )


def conversation_response(conv: Conversation, *, include_messages: bool = False) -> ConversationResponse:
    messages: list[MessageResponse] = []
    if include_messages:
        state = attributes.instance_state(conv)
        if "messages" not in state.unloaded:
            messages = [message_response(m) for m in conv.messages]
    return ConversationResponse(
        id=conv.id,
        title=conv.title,
        model=conv.model,
        dataset_enabled=bool(conv.dataset_enabled),
        dataset_ids=conv.dataset_ids,
        created_at=_as_utc(conv.created_at),
        updated_at=_as_utc(conv.updated_at),
        messages=messages,
    )


def user_profile_response(user: User) -> UserProfileResponse:
    return UserProfileResponse(
        id=user.id,
        phone=user.phone,
        nickname=user.nickname,
        is_verified=bool(user.is_verified),
        plan_type=enum_value(user.plan_type),
        total_tokens_used=int(user.total_tokens_used or 0),
        dataset_calls=int(user.dataset_calls or 0),
        daily_calls_used=int(user.daily_calls_used or 0),
        daily_call_limit=int(user.daily_call_limit or 0),
        created_at=_as_utc(user.created_at),
    )
